doc/advanced/content/executor_design.rst

```diff
@@ -0,0 +1,417 @@
+    };
+
+
+Best Fit Executor
+=================================
+
+In most scenarios users of PCL do not care much about performance and simply want their code to run
+fast without any additional steps performed from their side. For this reason their was a need for
+mechanism called best fit, which automatically selects best possible executor and its customizes its
+property in order to give the user good performance out of the box based on their system configurations
+and the function they are calling.
+
+The mechanism works be choosing an executor based on the availability of an executor (depends on
+hardware/software of a system) and the priority of executors specified in function by PCL
+maintainers & contributors which will give good performance. The executor properties can also be
+customized on for a better fit for certain scenarios. Besides this runtime checks are also be specified
+and on the basis of these runtime checks the executors are further filtered to selected most appropriate
+one. The two supported mechanisms in PCL currently are `enable_exec_with_priority` and
+`enable_exec_on_desc_priority`. You can read more about them in the code API.
+
+Alternative Designs Considered
+=================================
+
+There has been a lot of deliberation and discussion regarding all the design aspects of executors in PCL.
+The design has gone through multiple iterations before a consensus was reached. Some of the major alternate
+design proposals that were rejected were:
+
+* Tag Dispatching
+
+This was was one of the initial design considerations. It seemed like an attractive choice due to simplicity,
+could choose the allowed overload resolution to choose the best option at compile time by allowing
+tags to be inherited. These tags would serve as placeholders till executors were standardized.
+They lacked all the features of executors and offered no customization. Once executors were even standardized
+large parts of the codebase would need to be refactored to support executors.
+
+* Base Executor
+
+The idea was to have single base executor from which all executors would derive from. This would allow
+all the common cod to be shared among all the executors. The base executors has a CRTP based design which
+is used to access the properties of the derived executor. It also allowed simplification in many areas
+such as being able to reference any derived executor. Basic properties of all executors like
+copy constructors, overloaded equality operators.
+The CRTP mechanism had some restrictions in the sense that there was still a need for templates and
+their were concerns regarding explicitly passing all the properties to the base class using CRTP as
+it was felt to be unnecessary. Having inheritance also introduces runtime polymorphism which was against
+one of the design consideration i.e. have everything compile time and avoid any overhead.
+The base executor besides a few common functionality didn't do much and was discarded.
```

Comment:
The base executor didn't add a lot besides allowing code sharing of a few common functionality and thus it was discarded